### PR TITLE
Implement KafkaSchemaTransformReadConfiguration

### PR DESCRIPTION
**Pull request by damondouglas**
This PR address #21414 with a KafkaSchemaTransformReadConfiguration implementation.  It's design goals are to work with a KafkaSchemaTransformReadProvider that extends a [TypedSchemaTransformProvider](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/TypedSchemaTransformProvider.java).  Subsequent to this PR's approval/merge, the plan is to implement said corresponding  KafkaSchemaTransformReadProvider.

**Questions remain, however, how to configure the following in the setting of a [TypedSchemaTransformProvider](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/TypedSchemaTransformProvider.java):**

- [withCheckStopReadingFn](https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/io/kafka/KafkaIO.Read.html#withCheckStopReadingFn-org.apache.beam.sdk.transforms.SerializableFunction-)
- [withConsumerFactoryFn](https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/io/kafka/KafkaIO.Read.html#withConsumerFactoryFn-org.apache.beam.sdk.transforms.SerializableFunction-)
- [withKeyDeserializer](https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/io/kafka/KafkaIO.Read.html#withKeyDeserializer-java.lang.Class-)
- [withValueDeserializer](https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/io/kafka/KafkaIO.Read.html#withValueDeserializer-java.lang.Class-)

Due to historically failing beam_PreCommit_Java tests, to validate this PR, I ran the following prior to submission:

```
./gradlew rat
./gradlew spotlessCheck
./gradlew sdks:java:io:kafka:check
./gradlew sdks:java:io:kafka:checkStyleMain
```

I would like to request the following to review this PR:
R: @pabloem 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - ~Update `CHANGES.md` with noteworthy changes.~
 - ~If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).~

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by apilloud** [comment]
Looks like you requested @pabloem to review this, but he is out for another week. This really should get reviewed by someone working on IOs. cc: @johnjcasey

**Comment by johnjcasey** [comment]
Right now it looks like this is failing to compile, and spotless is also failing.

Can you address these?

**Comment by damondouglas** [comment]
> Right now it looks like this is failing to compile, and spotless is also failing.
> 
> Can you address these?
> Right now it looks like this is failing to compile, and spotless is also failing.
> 
> Can you address these?

Hello @johnjcasey Thank you for reviewing.

Are there other gradle commands you might recommend I run to detect where compile and spotless is failing?

I reran the following gradle tasks that already passed for me prior to submitting this PR.

```
./gradlew rat
./gradlew spotlessCheck
./gradlew sdks:java:io:kafka:check
./gradlew sdks:java:io:kafka:checkStyleMain
```

**Comment by pabloem** [comment]
I can review this this week. Thanks y'all

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaSchemaTransformReadConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+  @Nullable
+  public abstract Long getMaxNumRecords();
+
+  /**
+   * Reads a bounded amount of data from the unbounded Kafka topic resource. The bound is specified
+   * as an amount of time to read for. Each split of the source will read for this much time.
+   */
+  @Nullable
+  public abstract Long getMaxReadMillisecondsDuration();
+
```
Comment (pabloem) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaSchemaTransformReadConfiguration.java:84]: These are testing options so we can remove them I would say.

```suggestion
```

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaSchemaTransformReadConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+     */
+    public abstract Builder setMaxNumRecords(Long value);
+
+    /**
+     * Reads a bounded amount of data from the unbounded Kafka topic resource. The bound is
+     * specified as an amount of time to read for. Each split of the source will read for this much
+     * time.
+     */
+    public abstract Builder setMaxReadMillisecondsDuration(Long value);
+
```
Comment (pabloem) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaSchemaTransformReadConfiguration.java:156]: these are testing options so we can remove for now
```suggestion
```

**Comment by pabloem** [comment]
LGTM!

**Comment by pabloem** [comment]
Run Java PreCommit

**Comment by pabloem** [comment]
Run Java PreCommit

**Comment by pabloem** [comment]
thanks @damondouglas !

**Pull request merged**